python/modules/IcePy/Communicator.cpp

```diff
@@ -283,40 +284,31 @@ communicatorDestroy(CommunicatorObject* self, PyObject* /*args*/)
 }
 
 extern "C" PyObject*
-communicatorDestroyAsync(CommunicatorObject* self, PyObject* /*args*/)
+communicatorDestroyAsync(CommunicatorObject* self, PyObject* args)
 {
     assert(self->communicator);
-
     auto vfm = dynamic_pointer_cast<ValueFactoryManager>((*self->communicator)->getValueFactoryManager());
     assert(vfm);
 
-    auto futureType = reinterpret_cast<PyTypeObject*>(lookupType("Ice.Future"));
-    assert(futureType);
-
-    PyObjectHandle emptyArgs{PyTuple_New(0)};
-    PyObjectHandle future{futureType->tp_new(futureType, emptyArgs.get(), nullptr)};
-    if (!future.get())
+    PyObject* setResult;
+    if (!PyArg_ParseTuple(args, "O", &setResult))
     {
         return nullptr;
     }
 
-    // Call Ice.Future.__init__
-    futureType->tp_init(future.get(), emptyArgs.get(), nullptr);
-
-    // Create a new reference to prevent premature release of the future object. The reference will be released by
-    // the destroy callback.
-    PyObject* futureObject = Py_NewRef(future.get());
+    if (!PyCallable_Check(setResult))
+    {
+        PyErr_SetString(PyExc_TypeError, "setResult must be callable");
+        return nullptr;
+    }
 
     (*self->communicator)
         ->destroyAsync(
-            [self, vfm, futureObject]()
+            [self, setResult = Py_NewRef(setResult), vfm]()
```

Comment:
I don't understand why we need to make a new "ref" for `completed`.

This communicatorDestroyAsync is an internal function and I expect all callers (in our Python code) will ensure the completed callback remains alive until `completed` is called.

Reply:
Yes, you're right. Please add a comment.